src/log4net/Repository/Hierarchy/Hierarchy.cs

```diff
@@ -317,50 +203,23 @@ public override ILogger Exists(string name)
     public override ILogger[] GetCurrentLoggers()
     {
       // The accumulation in loggers is necessary because not all elements in
-      // ht are Logger objects as there might be some ProvisionNodes
-      // as well.
-      lock (m_ht)
-      {
-        System.Collections.ArrayList loggers = new System.Collections.ArrayList(m_ht.Count);
-
-        // Iterate through m_ht values
-        foreach (object node in m_ht.Values)
-        {
-          if (node is Logger)
-          {
-            loggers.Add(node);
-          }
-        }
-        return (Logger[])loggers.ToArray(typeof(Logger));
-      }
+      // loggers are Logger objects as there might be some ProvisionNodes as well.
+      return loggers.Values.OfType<Logger>().ToArray();
```

Comment:
Notice that the `Values`-property on ConcurrentDictionary is dangerous (locks the entire collection). Maybe do this instead:
```c#
return loggers.Select(l => l.Value).OfType<Logger>().ToArray();
```

Reply:
@snakefoot thanks for the hint - I was not aware of this.